x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/support/SamlAuthenticationState.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.idp.saml.support;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Represents a state object that can be used during an SP initiated flow in order to pass necessary parameters from the response of the
+ * /validate API to the /init API. State properties are not integrity protected, might end up in a cookie on the user's side and as such
+ * are treated as tainted parameters for informational purposes only. No authentication/authorization decisions should be made based on
+ * these.
+ */
+public class SamlAuthenticationState implements Writeable, ToXContentObject {
+    private String entityId;
+    private String requestedAcsUrl;
+    @Nullable
+    private String requestedNameidFormat;
+    @Nullable
+    private String authnRequestId;
```

Comment:
This ended up bothering me quite a bit but I raised the PR as is to discuss it. If this ends up in a cookie, a malicious user could forge the `inResponseTo` of  a SAML Response. This in itself is not a additional threat as it wouldn't allow any attack vectors that are not already mitigated by other means ( signature validation etc. ) 
The requestedNameIdFormat can also be forged _after_ the NameIdPolicy is validated but then again the only attack vector this enables is that a malicious user could i.e.have a persistent NameID generated for an SP that only needs/handles transient ones. 
Finally the entityId is used only for the SamlInitiateSingleSignOnResponse and could easiliy be even omitted , and the requestedAcsUrl can potentially be forged to send SAML Response messages to attacker controlled URLs. 

Although we could accept the risk these potential issues carry, it makes me think: Is this user facing state control the right trade off to satisfy the API design ( the split between /validate and /init ) and the interplay between them ? 

Is the alternative of having the /validate API require secondary authentication too and able to respond with SAML Responses worth it ?That would mean that the user needs to be authenticated _before_ the AuthnRequest will be parsed and validated. It would also mean that we would respond with authentication failure even if i.e. there is also an error and we can't satisfy the NameIdPolicy ( one might argue that in these cases we should reply with `InvalidNameIdPolicy` regardless of auhtn failure because even if the authentication was correct, we would _still_ fail )

Reply:
We discussed this in person and determined that we want to keep the distinction as is for the reasons that originally lead us to this design ( mainly extensibility and spec conformance ). 
The SamlAuthenticationState will not make it to a cookie for phase 1 and we will make it clear in the docs and code that this should be integrity protected if it ever needs to be shared in a format that is editable by the end user